core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -313,6 +303,33 @@ final class ShuffleBlockFetcherIterator(
   }
 }
 
+// Helper class that ensures a ManagedBuffer is released on InputStream.close()
+private class WrappedInputStream(
+              delegate: InputStream, buf: ManagedBuffer, var currentResult: FetchResult)
```

Comment:
Also, not that the original code would end up mutating the `currentResult` field in `ShuffleBlockFetcherIterator`, whereas the `currentResult = null` here only affects the field in this `WrappedInputStream` class and leaves the ShuffleBlockFetcherIterator's field untouched.